Source/WebCore/platform/mediastream/gstreamer/GStreamerDTMFSenderBackend.cpp

```diff
@@ -20,60 +20,197 @@
+    });
+
+    static uint32_t nPipeline = 0;
+    auto pipelineName = makeString("webkit-dtmf-playout-pipeline-"_s, nPipeline);
+    m_pipeline = gst_pipeline_new(pipelineName.ascii().data());
+    registerActivePipeline(m_pipeline);
+    connectSimpleBusMessageCallback(m_pipeline.get());
+    GST_DEBUG_OBJECT(m_pipeline.get(), "DTMF sender backend created");
+
+    auto audioconvert = makeGStreamerElement("audioconvert"_s);
+    auto audioresample = makeGStreamerElement("audioresample"_s);
+    auto queue = gst_element_factory_make("queue", nullptr);
+    auto sink = createPlatformAudioSink("dtmf"_s);
+    gst_bin_add_many(GST_BIN_CAST(m_pipeline.get()), m_dtmfSrc.get(), audioconvert, audioresample, queue, sink, nullptr);
+    gst_element_link_many(m_dtmfSrc.get(), audioconvert, audioresample, queue, sink, nullptr);
+    gst_element_set_state(m_pipeline.get(), GST_STATE_PLAYING);
 }
 
-bool GStreamerDTMFSenderBackend::canInsertDTMF()
+GStreamerDTMFSenderPrivate::~GStreamerDTMFSenderPrivate()
 {
-    notImplemented();
-    return false;
+    if (!m_pipeline)
+        return;
+
+    unregisterPipeline(m_pipeline);
+    disconnectSimpleBusMessageCallback(m_pipeline.get());
+    gst_element_set_state(m_pipeline.get(), GST_STATE_NULL);
 }
 
-void GStreamerDTMFSenderBackend::playTone(const char, size_t, size_t)
+struct StopData {
+    int tone;
+    GRefPtr<GstPad> pad;
+    ThreadSafeWeakPtr<GStreamerDTMFSenderPrivate> self;
+};
+WEBKIT_DEFINE_ASYNC_DATA_STRUCT(StopData);
+
+void GStreamerDTMFSenderPrivate::playTone(const RefPtr<RealtimeOutgoingAudioSourceGStreamer>& source, const char tone, size_t duration)
 {
-    notImplemented();
+    static HashMap<char, int, WTF::IntHash<char>, WTF::UnsignedWithZeroKeyHashTraits<char>> tones = {
+        { '0', 0 },
+        { '1', 1 },
+        { '2', 2 },
+        { '3', 3 },
+        { '4', 4 },
+        { '5', 5 },
+        { '6', 6 },
+        { '7', 7 },
+        { '8', 8 },
+        { '9', 9 },
+        { '*', 10 },
+        { '#', 11 },
+        { 'A', 12 },
+        { 'B', 13 },
+        { 'C', 14 },
+        { 'D', 15 }
+    };
+
+    auto element = adoptGRef(gst_bin_get_by_name(GST_BIN_CAST(source->bin().get()), "dtmfSource"));
+    auto pad = adoptGRef(gst_element_get_static_pad(element.get(), "src"));
+
+    if (tone == ',') {
+        GST_DEBUG_OBJECT(element.get(), "Playing silence for 2 seconds");
+        sleep(Seconds::fromMilliseconds(2000));
+        m_onTonePlayed();
+        GST_DEBUG_OBJECT(element.get(), "Playing tone %c DONE", tone);
+        return;
+    }
+
+    auto toneNumber = tones.get(tone);
+    GST_DEBUG_OBJECT(pad.get(), "Playing tone %c for %zu milliseconds", tone, duration);
+    sendEvent(pad, toneNumber, 25, true);
+
+    auto data = createStopData();
+    data->tone = toneNumber;
+    data->pad = WTFMove(pad);
+    data->self = this;
+
+    g_timeout_add_full(G_PRIORITY_DEFAULT, duration, reinterpret_cast<GSourceFunc>(+[](gpointer userData) -> gboolean {
```

Comment:
Why can't you just use webkit function to run on main thread?

Reply:
That's what came to me first, but yes, a one shot timer should also work.